src/librbd/mirror/snapshot/CreatePrimaryRequest2.h

```diff
@@ -0,0 +1,152 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MIRROR_SNAPSHOT_CREATE_PRIMARY_REQUEST2_H
+#define CEPH_LIBRBD_MIRROR_SNAPSHOT_CREATE_PRIMARY_REQUEST2_H
+
+#include "include/buffer.h"
+#include "include/rados/librados.hpp"
+#include "cls/rbd/cls_rbd_types.h"
+#include "librbd/internal.h"
+#include "librbd/mirror/snapshot/Types.h"
+
+#include <string>
+#include <set>
+
+struct Context;
+
+namespace librbd {
+
+struct ImageCtx;
+
+namespace mirror {
+namespace snapshot {
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class CreatePrimaryRequest2 {
```

Comment:
@nbalacha 
At a high level can you state the objective of CreatePrimaryRequest2?

Is it a state machine specifically to create primary snapshots of a group's member images such that the each async operation proceeds only after it completes for all the member images? And this is different from the current model of calling the **entire set** of async operations of group member images concurrently?

Reply:
That is correct. CreatePrimaryRequest2 will perform each operation on every image in the vector before proceeding to the next. This is differs from the current model where, in case of a failure, there is no way to know how far CreatePrimary request got for each image.
With the approach taken in CreatePrimaryRequest2, higher level group operations will know exactly which stage failed and tailor the cleanup accordingly.

Reply:
Please note that I have not tested the error code paths for CreatePrimaryRequest2.